### DISCUSSION TITLE
Using URL parameters to record participant IDs from a survey company

### DISCUSSION THREAD

**Discussion by willking98** [surveydown-dev/surveydown#92] [Feature Requests]
I am wondering whether the potential to include URL parameter parsing in the `server.R` function is possible. Having recently used Prolific to recruit participants without using URL parameters I would definitely recommend using this feature if possible rather than allowing participants to input their own `userID`. 

Below is some code that I think should solve this, though is not functional as session object is not found.  

```
query <- parseQueryString(session$clientData$url_search)

prolific_id <- if (!is.null(query[['PROLIFIC_PID']])) {
  query[['PROLIFIC_PID']]
} else {
  "ID:noURLparam"
}

sd_store_value(prolificID)

```

**Comment by jhelvy** [comment]
We literally just added this feature over the past few days :)
Still need to update the documentation on it, but as of [v0.2.2](https://pkg.surveydown.org/news/#surveydown-022) we now have a [`sd_get_url_pars()`](https://pkg.surveydown.org/reference/sd_get_url_pars) function to reactively get the parameters in the URL, and a [`sd_redirect()`](https://pkg.surveydown.org/reference/sd_redirect) function to allow you to send respondents to another site at the end of the survey, either a static url or a reactive one that can be created in the server chunk and have the scrapped URL parameters included.

@pingfan-hu made a demo survey [here](https://github.com/surveydown-dev/demo-external-redirect), and he's still working on a video to showcase it. It will eventually be posted on the [demos page](https://surveydown.org/demos), and we'll add a whole article on it in the documentation.

**Reply by willking98** [reply]
Fantastic! This is great news. Well done again!

In terms of the redirect, I have been using some JS code to display a static url on the last page (below) but that's great to hear that you're sd_redirect will be along soon for reactive urls. Looking forward to the documentation!

```
::: {#end_normal .sd-page}

# End Page

# start chunk {r}
completion_code <- sample(10^5:(10^6 - 1), 1)
# end chunk 

The survey is finished. Thank you for your feedback!

Please click the button below to complete your participation 
<br> 
 
<br>
<a href="https://www.dynata.com" target="_blank" class="btn btn-primary">Return to Dynata</a>.
<br>

Completion code: **`r completion_code`**.

You may close the window now.
:::

**Comment by jhelvy** [comment]
This could also work if the respondent is able to provide a completion code, but a few points:

When you make a completion code in the main body like this:

```r
completion_code <- sample(10^5:(10^6 - 1), 1)
```

This will create a single code that is the same for all respondents. It's basically just running once and then getting embedded in the survey. Also, this code doesn't get stored in your response data, so you won't know what the code is unless you go to the end of the survey yourself to get it.

If instead you wanted to generate a unique code for each respondent, then you should create the code [reactively](https://surveydown.org/reactivity) in the server code chunk. So your server chunk would have something like this somewhere before the main call to `sd_server()`:

```r
completion_code <- sample(10^5:(10^6 - 1), 1)
sd_store_value(completion_code, id = 'completion_code') # Can be any id you want
```

This will store the value in your response data so you'll know the code for each respondent. Then in your survey you could display the created code like this:

````qmd
Completion code: **`r sd_output('completion_code')`**
````

We recently made the `sd_output()` function for displaying reactive values (things generated in the server code chunk, or things like values to previously answered questions, etc.

**Reply by jhelvy** [reply]
I wrote this but never actually tested it before. I just tried today and noticed that the completion code wasn't actually showing in the survey with the code I wrote above. I just pushed a fix for this so now you can display the code like this:

```qmd
Completion code: **`r sd_output('completion_code', type = 'value')`**
```

The `type = 'value'` part was missing before and is needed so `sd_output()` knows where to look for the values. The goal is for anything value created using an `sd_*()` function (e.g. `sd_question()`, `sd_store_value()`, etc.) should be accessible with `type = 'value'` in `sd_output()`.

**Reply by jhelvy** [reply]
Also btw, we added a `sd_completion_code()` function to make it easier to generate these. See [here](https://surveydown.org/reactivity#displaying-stored-values-e.g.-a-completion-code).

**Comment by jhelvy** [comment]
I'm going to close this as this is now supported.
